File: fragments/solana_airdrop/solana_airdrop_utils.go
package solana_airdrop

import (
	"ff/solana_rpc"
	solana_transaction "ff/solana_transaction"

	"context"
	"fmt"

	"github.com/dustin/go-humanize"
	"github.com/gagliardetto/solana-go"
	"github.com/gagliardetto/solana-go/rpc"
)

func SendAndConfirmAirdrop(address solana.PublicKey, amount uint64) error {
	client := solana_rpc.InitRpcClient()

	fmt.Printf("Airdropping %s lamports to %s\n", humanize.Comma(int64(amount)), address)

	sig, err := client.RequestAirdrop(context.Background(), address, amount, rpc.CommitmentConfirmed)
	if err != nil {
		return fmt.Errorf("failed to request airdrop: %w", err)
	}

	err = solana_transaction.ConfirmRecentTransaction(sig)
	if err != nil {
		return fmt.Errorf("failed to confirm airdrop: %w", err)
	}

	fmt.Printf("Airdrop confirmed for %s with signature: %s\n", address, sig)
	return nil
}

File: fragments/solana_airdrop/solana_airdrop_utils.rs
use crate::solana_rpc::solana_rpc_utils::init_rpc_client;
use crate::solana_transaction::solana_transaction_utils::confirm_recent_signature;
use num_format::{Locale, ToFormattedString};
use solana_client::client_error::{ClientError, ClientErrorKind};
use solana_sdk::pubkey::Pubkey;

pub fn send_and_confirm_airdrop(pubkey: Pubkey, amount: u64) -> Result<(), ClientError> {
    let client = init_rpc_client();

    println!(
        "Airdropping {} lamports to {}",
        amount.to_formatted_string(&Locale::en_GB),
        pubkey
    );

    let signature = client.request_airdrop(&pubkey, amount)?;
    let is_confirmed = confirm_recent_signature(&signature, None)?;

    if !is_confirmed {
        return Err(ClientError {
            request: Some(solana_client::rpc_request::RpcRequest::Custom {
                method: "ConfirmTransaction",
            }),
            kind: ClientErrorKind::Custom(
                "Airdrop transaction not confirmed within the timeout period".to_string(),
            ),
        });
    }

    println!(
        "Airdrop confirmed for {} with signature {}",
        pubkey, signature
    );

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use solana_sdk::{
        commitment_config::CommitmentConfig, native_token::LAMPORTS_PER_SOL, signature::Keypair,
        signer::Signer,
    };

    #[test]
    fn test_solana_airdrop() {
        let address = Keypair::new().pubkey();
        let rpc_client = init_rpc_client();
        let commitment = CommitmentConfig::confirmed();
        let airdrop_amount = LAMPORTS_PER_SOL;
        let _ = send_and_confirm_airdrop(address, airdrop_amount);

        let balance = rpc_client
            .wait_for_balance_with_commitment(&address, Some(airdrop_amount), commitment)
            .unwrap();

        assert_eq!(balance, airdrop_amount);
    }
}
